File: 2024/2/day2.js
const fs = require("fs");

const input = fs.readFileSync("./inputs/input.txt", "utf8");

const sequences = input.split("\n").map((line) => line.split(" ").map(Number));

const isValidSequence = (numbers) => {
  if (numbers.length < 2) return false;

  const diffs = numbers.slice(1).map((n, i) => n - numbers[i]);

  const validDiffs = diffs.every(
    (diff) => Math.abs(diff) >= 1 && Math.abs(diff) <= 3
  );

  const allPositive = diffs.every((diff) => diff > 0);
  const allNegative = diffs.every((diff) => diff < 0);

  return validDiffs && (allPositive || allNegative);
};

const isValidSequenceWithOneError = (numbers) => {
  for (let i = 0; i < numbers.length; i++) {
    const newNumbers = [...numbers];
    newNumbers.splice(i, 1);
    if (isValidSequence(newNumbers)) return true;
  }
  return false;
};

const validSequences = sequences.filter(isValidSequence).length;
const validSequencesWithOneError = sequences.filter(
  isValidSequenceWithOneError
).length;

console.log(validSequences);
console.log(validSequencesWithOneError);

File: 2024/2/day2.py
input = open("./inputs/input.txt", "r").read()

sequences = [list(map(int, line.split(" "))) for line in input.split("\n")]


def is_valid_sequence(numbers):
    if len(numbers) < 2:
        return False

    diffs = [numbers[i] - numbers[i + 1] for i in range(len(numbers) - 1)]
    
    valid_diffs = all(1 <= abs(diff) <= 3 for diff in diffs)
    all_positive = all(diff > 0 for diff in diffs)
    all_negative = all(diff < 0 for diff in diffs)

    return valid_diffs and (all_positive or all_negative)

def is_valid_sequence_with_one_error(numbers):
    for i in range(len(numbers)):
        new_numbers = numbers[:i] + numbers[i+1:]
        if is_valid_sequence(new_numbers):
            return True
    return False

valid_sequences = sum(1 for sequence in sequences if is_valid_sequence(sequence))
valid_sequences_with_one_error = sum(1 for sequence in sequences if is_valid_sequence_with_one_error(sequence))

print(valid_sequences)
print(valid_sequences_with_one_error)
